src/tir/schedule/primitive/fuse_split.cc

```diff
@@ -0,0 +1,483 @@
+  // order with before.
+  // Step 1. Check correctness
+  GetScopeRootAndCheckStagePipeline(self, loop_sref);
+  const auto* loop = loop_sref->StmtAs<ForNode>();
+  if (loop == nullptr) {
+    throw NotLoopError(self->mod, loop_sref->stmt->GetTypeKey());
+  }
+  if (!loop->annotations.empty()) {
+    throw HasAnnotationError(self->mod, GetRef<For>(loop));
+  }
+  if (loop->thread_binding.defined()) {
+    throw HasThreadBindingError(self->mod, GetRef<For>(loop));
+  }
+  // Currently, loops starting with 0 is not supported
+  arith::Analyzer analyzer;
+  if (!analyzer.CanProve(loop->min == 0)) {
+    throw LoopNotStartWithZeroError(self->mod, GetRef<For>(loop));
+  }
+  PrimExpr tot_length = 1;
+  int infer_index = -1;
+  for (size_t i = 0; i < factors.size(); i++) {
+    if (!analyzer.CanProve(factors[i] == -1)) {
+      tot_length *= factors[i];
+    } else {
+      if (infer_index != -1) {
+        throw NotSingleInferFactorError(self->mod);
+      } else {
+        infer_index = i;
+      }
+    }
+  }
+  // Step 2. infer factors if needed
+  Array<PrimExpr> inferred_factors(factors);
+  if (infer_index != -1) {
+    inferred_factors.Set(infer_index,
+                         analyzer.Simplify(floordiv(loop->extent + tot_length - 1, tot_length)));
+  } else {
+    if (!analyzer.CanProve(tot_length >= loop->extent)) {
+      throw WrongFactorProductError(self->mod, GetRef<For>(loop));
+    }
+  }
+  // Step 3. Replace all occurrence of the original loop var with new variables
+  std::vector<Var> new_loop_vars;
+  new_loop_vars.reserve(inferred_factors.size());
+  for (size_t i = 0; i < inferred_factors.size(); i++) {
+    new_loop_vars.push_back(loop->loop_var.copy_with_suffix("_" + std::to_string(i)));
+  }
+  PrimExpr substitute_value = 0;
+  for (size_t i = 0; i < inferred_factors.size(); i++) {
+    substitute_value *= inferred_factors[i];
+    substitute_value += new_loop_vars[i];
+  }
+  Map<Block, Block> opaque_block_reuse;
+  auto substitute_function = [&](const Var& v) -> Optional<PrimExpr> {
+    if (v.same_as(loop->loop_var)) {
+      return substitute_value;
+    } else {
+      return NullOpt;
+    }
+  };
+  Stmt new_loop_body =
+      SubstituteAndCollectOpaqueBlock(loop->body, &opaque_block_reuse, substitute_function);
+  for (size_t i = 0; i < inferred_factors.size(); i++) {
+    analyzer.Bind(new_loop_vars[i], Range::FromMinExtent(0, inferred_factors[i]));
+  }
+  // Step 4. Update predicate to guard the loop
+  PrimExpr predicate = substitute_value < loop->extent;
+  new_loop_body = PredicateUpdater(predicate, &analyzer)(new_loop_body);
+  // Step 5. Generate tnested loops to replace the original loop and simplify the binding
+  Stmt outer_stmt = new_loop_body;
+  for (int i = inferred_factors.size() - 1; i >= 0; i--) {
+    outer_stmt = For(new_loop_vars[i], 0, inferred_factors[i], loop->kind, outer_stmt);
+  }
+
+  outer_stmt =
+      Downcast<For>(SimplifyBindings(outer_stmt, GetLoops(loop_sref), &opaque_block_reuse));
```

Comment:
would you like to give an example where we need to simplify the bindings?

Reply:
I think it's not needed. In most cases we'll need to simplify the bindings because some nested expressions will be generated after substitution.